### ISSUE TITLE
Build failure with assert()

### ISSUE THREAD

**Issue by yan12125** [mozilla/rr#2000]
On Arch Linux, with GCC 6.3.1 and glibc 2.25:
```
[  7%] Building C object CMakeFiles/ptrace_remote_unmap.dir/src/test/ptrace_remote_unmap.c.o
/home/yen/Projects/tmp/rr/src/test/ptrace_remote_unmap.c: In function ‘munmap_remote’:
/home/yen/Projects/tmp/rr/src/test/ptrace_remote_unmap.c:53:3: error: suggest parentheses around assignment used as truth value [-Werror=parentheses]
   assert(wret = child);
   ^~~~~~
/home/yen/Projects/tmp/rr/src/test/ptrace_remote_unmap.c:59:3: error: suggest parentheses around assignment used as truth value [-Werror=parentheses]
   assert(wret = child);
   ^~~~~~
/home/yen/Projects/tmp/rr/src/test/ptrace_remote_unmap.c: In function ‘main’:
/home/yen/Projects/tmp/rr/src/test/ptrace_remote_unmap.c:117:3: error: suggest parentheses around assignment used as truth value [-Werror=parentheses]
   assert(wret = child);
   ^~~~~~
/home/yen/Projects/tmp/rr/src/test/ptrace_remote_unmap.c:123:3: error: suggest parentheses around assignment used as truth value [-Werror=parentheses]
   assert(wret = child);
   ^~~~~~
cc1: all warnings being treated as errors
make[2]: *** [CMakeFiles/ptrace_remote_unmap.dir/build.make:63: CMakeFiles/ptrace_remote_unmap.dir/src/test/ptrace_remote_unmap.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1358: CMakeFiles/ptrace_remote_unmap.dir/all] Error 2
make: *** [Makefile:161: all] Error 2
```

The relevant #define in /usr/include/assert.h:
```C
/* When possible, define assert so that it does not add extra
   parentheses around EXPR.  Otherwise, those added parentheses would
   suppress warnings we'd expect to be detected by gcc's -Wparentheses.  */
# if !defined __GNUC__ || defined __STRICT_ANSI__
#  define assert(expr)							\
    ((expr)								\
     ? __ASSERT_VOID_CAST (0)						\
     : __assert_fail (#expr, __FILE__, __LINE__, __ASSERT_FUNCTION))
# else
#  define assert(expr)							\
    ({									\
      if (expr)								\
        ; /* empty */							\
      else								\
        __assert_fail (#expr, __FILE__, __LINE__, __ASSERT_FUNCTION);	\
    })
# endif
```

The following fix works for me:
```Diff
diff --git a/src/test/ptrace_remote_unmap.c b/src/test/ptrace_remote_unmap.c
index 8f4c7eac..5d14f15a 100644
--- a/src/test/ptrace_remote_unmap.c
+++ b/src/test/ptrace_remote_unmap.c
@@ -50,13 +50,13 @@ void munmap_remote(pid_t child, uintptr_t start, size_t size) {
   checked_ptrace(PTRACE_SYSCALL, child, 0, 0);
   // Wait until entry trap
   wret = waitpid(child, &status, __WALL | WSTOPPED);
-  assert(wret = child);
+  assert((wret = child));
   assert(WSTOPSIG(status) == (SIGTRAP | 0x80));
 
   checked_ptrace(PTRACE_SYSCALL, child, 0, 0);
   // Wait until exit trap
   wret = waitpid(child, &status, __WALL | WSTOPPED);
-  assert(wret = child);
+  assert((wret = child));
   assert(WSTOPSIG(status) == (SIGTRAP | 0x80));
   // Verify that the syscall didn't fail
   checked_ptrace(PTRACE_GETREGSET, child, (void*)NT_PRSTATUS, &iov);
@@ -114,13 +114,13 @@ int main(void) {
 
   // That caused another stop
   wret = waitpid(child, &status, __WALL | WSTOPPED);
-  assert(wret = child);
+  assert((wret = child));
 
   // Continue until the exec
   checked_ptrace(PTRACE_CONT, child, 0, 0);
   // This should be the exec stop
   wret = waitpid(child, &status, __WALL | WSTOPPED);
-  assert(wret = child);
+  assert((wret = child));
   assert(status >> 8 == (SIGTRAP | (PTRACE_EVENT_EXEC << 8)));
 
   // On kernels with aggressive ASLR, the executable mapping may
```
From preprocessor outputs the ```if(expr)``` approach is chosen. Comments in assert.h say that it omits parantheses deliberately. As a result, I'm not sure whether my fix is correct or not.

**Comment by janisozaur** [comment]
The problem here is you never want to execute code in asserts, because generally in release mode (`-DNDEBUG`) they get removed and the code is not executed. Adding parentheses would suppress this compiler warning which is here precisely for the reason I stated above.

If you look at this file closer, you'll also notice there are other places where the condition is written correctly: https://github.com/mozilla/rr/blob/68d9aab/src/test/ptrace_remote_unmap.c#L108

**Comment by Keno** [comment]
Actually I think those assignments are wrong and should be comparisons instead. Also `test_assert` instead of `assert`.

**Comment by janisozaur** [comment]
@Keno already fixed all that in #2001 :)
